kbase-extension/static/kbase/js/widgets/narrative_core/kbaseNarrativeOutlinePanel.js

```diff
@@ -0,0 +1,260 @@
+            const stack = [root_node];
+            outline_items.forEach((item) => {
+                while (stack[stack.length - 1].item.depth >= item.depth) {
+                    stack.pop();
+                }
+                const node = {
+                    item: item,
+                    children: []
+                };
+                stack[stack.length - 1].children.push(node);
+                stack.push(node);
+            });
+            return root_node;
+        },
+
+        /** 
+         * Render a nested outline node and its children recursively 
+         * @param {OutlineNode} node - The node to render
+         * @returns {jQuery} - The rendered node
+        */
+        renderOutlineNode: function (node) {
+            // If not the root node, create the item
+            if (node.item.depth !== 0) {
+                node.content = $('<div>')
+                    .addClass('kb-narr-outline__item')
+                    .attr('title', node.item.title)
+                    .append(
+                        $('<span>')
+                            .addClass('kb-narr-outline__item-icon-wrapper')
+                            .append($(node.item.icon).addClass('kb-narr-outline__item-icon'))
+                    )
+                    .append(
+                        $('<a>')
+                            .text(node.item.title)
+                            .attr('href', '#')
+                            .attr('title', node.item.title)
+                            .addClass('kb-narr-outline__item-content')
+                            .click(() => this.scrollToItem(node.item))
+                    )
+                    .click(() => this.scrollToItem(node.item));
+            }
+
+            // Render the children
+            const children = $('<ul>');
+            node.children.forEach((child) => {
+                children.append(this.renderOutlineNode(child));
+            });
+
+            if (node.item.depth === 0) {
+                return children;
+            } else {
+                return $('<li>').append(node.content).append(children);
+            }
+        },
+
+        /**
+         * Scroll to the cell corresponding to the given outline item
+         * @param {OutlineItem} item - The outline item to scroll to
+         */
+        scrollToItem: function (item) {
+            const nb = $('#notebook-container');
+            const scroll_to = nb.scrollTop() + $(item.element).offset().top;
+            nb.get(0).scrollTo({
+                // scroll further down to account for header
+                top: scroll_to - this.options.headerHeight,
+                behavior: 'smooth',
+            });
+            item.element.click(); // select the cell
+        },
+
+        /**
+         * Check if the given cell element is visible on screen, accounting for header height
+         * @param {HTMLElement} el - The cell element to check
+         */
+        cellInView: function (el) {
```

Comment:
This is similar to a utility function in `kbase-extension/static/kbase/js/util/display.js` (`verticalInViewport`). I think it would be helpful to have this be there, and also be available for other things that need it, like lazy loading, which is implemented slightly differently in a few places.

Reply:
It's definitely similar, though, this needs to account for the header height, which the utility does not (though, I suppose it _should_) so I'll take a look at wether changing that has side effects